diff --git a/ytfzf b/ytfzf
index ef79374..7c12ab6 100755
--- a/ytfzf
+++ b/ytfzf
@@ -141,7 +141,7 @@ trim_url () {
 }
 
 command_exists "quick_menu" || quick_menu () {
-	fzf --reverse --prompt="$1"
+	fzf --prompt="$1"
 }
 command_exists "quick_menu_ext" || quick_menu_ext (){
 	external_menu "$1"
@@ -1182,7 +1182,7 @@ interface_text () {
 			video_info_text
 		done |
 		column -t -s "$tab_space" |
-		fzf -m --tabstop=1 --layout=reverse --expect="$shortcut_binds" | set_keypress |
+		fzf | set_keypress |
 		trim_url > "$selected_id_file"
 	#we can't do handle_keypress < "$keypress_file" because it leaves fd0 open which breaks $search_again_shortcut
 	handle_keypress "$(cat "$keypress_file")"
@@ -1446,9 +1446,9 @@ interface_thumbnails () {
 	#fzf_preview_side will get reset if we don't pass it in
 	jq -r '.[]|[.title,"'"$gap_space"'|"+.channel,"|"+.duration,"|"+.views,"|"+.date,"|"+.url]|@tsv' < "$video_json_file" |
 	sort_video_data_fn |
-	fzf -m \
-	--preview "__is_fzf_preview=1 fzf_preview_side='$fzf_preview_side' scrape='$scrape' thumb_dir='$thumb_dir' YTFZF_PID='$YTFZF_PID' UEBERZUG_FIFO='$UEBERZUG_FIFO' $0 -U preview_img '$thumbnail_viewer' {} '$video_json_file'" \
-	--preview-window "$fzf_preview_side:50%:wrap" --layout=reverse --expect="$shortcut_binds" | set_keypress |
+	fzf \
+	"__is_fzf_preview=1 fzf_preview_side='$fzf_preview_side' scrape='$scrape' thumb_dir='$thumb_dir' YTFZF_PID='$YTFZF_PID' UEBERZUG_FIFO='$UEBERZUG_FIFO' $0 -U preview_img '$thumbnail_viewer' {} '$video_json_file'" \
+	"$fzf_preview_side:50%:wrap" | set_keypress |
 	trim_url > "$selected_id_file"
 
 	preview_stop "$thumbnail_viewer"
-- 
2.35.1

