--- ./output/build.ninja
+++ ./output/build.ninja
@@ -101,8 +101,8 @@
  DEPFILE_UNQUOTED = wbg.p/jpg.c.o.d
  ARGS = -Iwbg.p -I. -I.. -Isubprojects/tllist -I../subprojects/tllist -I/usr/include/pixman-1 -I/usr/include/libpng16 -fdiagnostics-color=always -DNDEBUG -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -Werror -std=c18 -O3 -D_POSIX_C_SOURCE=200809L -D_GNU_SOURCE -fmacro-prefix-map=../= -DWBG_HAVE_PNG=1 -DWBG_HAVE_JPG=1 -O2 -pipe -march=znver1 -pthread
 
-build wbg: c_LINKER wbg.p/meson-generated_.._wlr-layer-shell-unstable-v1.c.o wbg.p/meson-generated_.._xdg-shell.c.o wbg.p/main.c.o wbg.p/log.c.o wbg.p/shm.c.o wbg.p/png.c.o wbg.p/jpg.c.o | /usr/lib/../lib/libffi.so /usr/lib/../lib/libwayland-client.a /usr/lib/libjpeg.so /usr/lib/libm.a /usr/lib/libpixman-1.so /usr/lib/libpng16.so /usr/lib/libz.so
- LINK_ARGS = -Wl,--as-needed -Wl,--no-undefined -Wl,-O1 -O2 -pipe -march=znver1 -Wl,--start-group /usr/lib/libpixman-1.so /usr/lib/libpng16.so /usr/lib/libz.so /usr/lib/libjpeg.so /usr/lib/../lib/libwayland-client.a -lm -pthread /usr/lib/../lib/libffi.so -Wl,--end-group
+build wbg: c_LINKER wbg.p/meson-generated_.._wlr-layer-shell-unstable-v1.c.o wbg.p/meson-generated_.._xdg-shell.c.o wbg.p/main.c.o wbg.p/log.c.o wbg.p/shm.c.o wbg.p/png.c.o wbg.p/jpg.c.o | /usr/lib/../lib/libffi.a /usr/lib/../lib/libwayland-client.a /usr/lib/libjpeg.a /usr/lib/libm.a /usr/lib/libpixman-1.a /usr/lib/libpng16.a /usr/lib/libz.a
+ LINK_ARGS = -Wl,--as-needed -Wl,--no-undefined -static -O2 -pipe -march=znver1 -Wl,--start-group /usr/lib/libpixman-1.a /usr/lib/libpng16.a /usr/lib/libz.a /usr/lib/libjpeg.a /usr/lib/../lib/libwayland-client.a -lm -pthread /usr/lib/../lib/libffi.a -Wl,--end-group
 
 # Test rules
 
